clientapi/routing/room_tagging.go

```diff
@@ -0,0 +1,265 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package routing
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/matrix-org/dendrite/clientapi/auth/authtypes"
+	"github.com/matrix-org/dendrite/clientapi/auth/storage/accounts"
+	"github.com/matrix-org/dendrite/clientapi/httputil"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/dendrite/clientapi/producers"
+	"github.com/matrix-org/gomatrix"
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/matrix-org/util"
+)
+
+// newTag creates and returns a new Tag type
+func newTag() gomatrix.TagContent {
+	return gomatrix.TagContent{
+		Tags: make(map[string]gomatrix.TagProperties),
+	}
+}
+
+// GetTags implements GET /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags
+func GetTags(
+	req *http.Request,
+	accountDB *accounts.Database,
+	device *authtypes.Device,
+	userID string,
+	roomID string,
+	syncProducer *producers.SyncAPIProducer,
+) util.JSONResponse {
+
+	if device.UserID != userID {
+		return util.JSONResponse{
+			Code: http.StatusForbidden,
+			JSON: jsonerror.Forbidden("Cannot retrieve another user's tags"),
+		}
+	}
+
+	_, data, err := obtainSavedTags(req, userID, roomID, accountDB)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	dataByte, err := json.Marshal(data)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	var tagData []gomatrixserverlib.ClientEvent
+	tagContent := newTag()
+	err = json.Unmarshal(dataByte, &tagData)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	err = json.Unmarshal(tagData[0].Content, &tagContent)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	// send data to syncapi
+	if err := syncProducer.SendData(userID, roomID, "m.tag"); err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: tagContent,
+	}
+}
+
+// PutTag implements PUT /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags/{tag}
+// Put functionality works by getting existing data from the DB (if any), adding
+// the tag onto the "map" and saving the new "map" onto the DB
+func PutTag(
+	req *http.Request,
+	accountDB *accounts.Database,
+	device *authtypes.Device,
+	userID string,
+	roomID string,
+	tag string,
+	syncProducer *producers.SyncAPIProducer,
+) util.JSONResponse {
+
+	if device.UserID != userID {
+		return util.JSONResponse{
+			Code: http.StatusForbidden,
+			JSON: jsonerror.Forbidden("Cannot modify another user's tags"),
+		}
+	}
+
+	localpart, data, err := obtainSavedTags(req, userID, roomID, accountDB)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	var properties gomatrix.TagProperties
+
+	if reqErr := httputil.UnmarshalJSONRequest(req, &properties); reqErr != nil {
+		return *reqErr
+	}
+
+	tagContent := newTag()
+	var dataByte []byte
+	if len(data) > 0 {
+		dataByte, err = json.Marshal(data)
```

Comment:
Since you do this Marshal/Unmarshal data dance a few times in this PR, it may be worth breaking it out into its own function with an explanation on how it works?

Reply:
Sure!